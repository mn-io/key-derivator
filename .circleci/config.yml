version: 2
jobs:
  build:
    docker:
      - image: circleci/node:12.2.0
    steps:
      - checkout
      - run:
          name: Setup Dependencies
          command: npm install
      - run:
          name: Setup Code Climate test-reporter
          command: |
            curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter
            chmod +x ./cc-test-reporter
      - run: # run tests
          name: Run Test and Coverage
          command: |
            ./cc-test-reporter before-build
            npm test -- --coverage
            ./cc-test-reporter after-build --exit-code $?