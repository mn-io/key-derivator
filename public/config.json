{
  "hashBox": {
    "tokenSalt": "fdF6e%! #wMe",
    "keySalt": "f134§",

    "tokenHashingIterations": 6500,
    "rowHashIterations": 65,

    "hashResultLengthInBytes": 128,

    "outputRows": 24,
    "outputColumns": [12, 24, -1],
    "validCharacters": "abcdefghijklmnopqrstuvwxyz ABCDEFGHIJKLMNOPQRSTUVWXYZ äöüÄÖÜ 1234567890 abcdefghijklmnopqrstuvwxyz ABCDEFGHIJKLMNOPQRSTUVWXYZ äöüÄÖÜ 1234567890 \"!$%&(){}[]-_.:,;#+*@"
  },

  "profiles": {
    "Skype": {
      "validCharacters": "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZäöüÄÖÜ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZäöüÄÖÜ1234567890!\"#$%&,!\"#$%&,",
      "constraints": ["(?=(.*\\d){2})", "(?=.*[a-zA-Z]){2}"]
    }
  },

  "UI": {
    "colors": ["#3F51B5", "#2196F3", "#03A9F4", "#00BCD4", "#009688", "#4CAF50", "#8BC34A", "#CDDC39", "#FFEB3B", "#FFC107", "#FF9800", "#FF5722", "#F44336", "#E91E63", "#9C27B0", "#673AB7"]
  }
}
