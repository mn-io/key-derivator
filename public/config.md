# Constraints
 * No beginning whitespace: `(^\\S).*`
 * No trailing whitespace:`.*\\S$`
 * Min 1 Number: `(?=(.*\\d){1})`
 * Min 2 Letters: `(?=.*[a-zA-Z]){2}`
 * Min 3 of those characters: `(?=.*[!@#$~-]){3}`
