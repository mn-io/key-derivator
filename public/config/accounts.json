{
  "acdasdcasdc": {
      "type": "skype",
      "salt": "asdcadfvdf124234"
    }
}
