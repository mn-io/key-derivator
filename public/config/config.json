{
  "globalSalt": "asdcasdhckdscluh",
  "fieldCount": 3,
  "profiles": {
    "skype": {
      "specialChars": ["#", "$"]
    }
  } 
}
