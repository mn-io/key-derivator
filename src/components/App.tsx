import Paper from '@material-ui/core/Paper'
import { Theme } from '@material-ui/core/styles/createMuiTheme'
import { StyleRules } from '@material-ui/core/styles/withStyles'
import withStyles from '@material-ui/core/styles/withStyles'
import * as React from 'react'
import IConfig, { IProfile } from '../@types/Config'
import { WithStyleProps } from '../@types/Styles'
import { ConfigProvider } from '../services/ConfigProvider'
import emitter from '../services/Emitter'
import HashBox from '../services/HashBox'
import { Logger } from '../services/Logger'
import HashOutput from './HashOutput'
import UserInput from './UserInput'

const configUrl = 'config.json'

type Props = WithStyleProps

type State = {
  config?: IConfig,
  inputIdentifier?: string,
  inputToken?: string,
  profile?: IProfile,
  hashes?: string[][],
}

const styles = (theme: Theme) => ({
  cell: {
    display: 'table-cell',
    verticalAlign: 'top',
  },
  hashOutput: {
    marginLeft: theme.spacing.unit * 2,
    padding: theme.spacing.unit * 2,
  },
  root: {
    display: 'table',
    margin: 'auto',
    marginTop: theme.spacing.unit * 2,
    width: 'auto',
  },
  row: {
    display: 'table-row',
  },
  userInput: {
    padding: theme.spacing.unit * 2,
    width: 250,
  },
}) as StyleRules

class App extends React.Component<Props, State> {

  private previousInput = ''
  private previousProfile?: IProfile
  private mounted = false

  constructor(props: Props) {
    super(props)

    this.loadConfig()
    Logger.interceptConsole(emitter)
    void HashBox.selfTest()

    this.state = {}
  }

  public render() {
    const { config, hashes } = this.state
    const { classes } = this.props

    return (
      <div className={classes.root}>
        <div className={classes.row}>
          <div className={classes.cell}>
            <Paper className={classes.userInput}>
              <UserInput config={config} onGenerateHashes={this.onGenerateHashes} />
            </Paper>
          </div>
          <div className={classes.cell}>
            {
              config && hashes &&
              <Paper className={classes.hashOutput}>
                <HashOutput colors={config.ui.colors} otm={config.otm} hashes={hashes} />
              </Paper>
            }
          </div>
        </div>
      </div>
    )
  }

  public componentDidMount() {
    this.mounted = true
  }

  public componentWillUnmount() {
    this.mounted = false
  }

  private loadConfig = async () => {
    const config = await ConfigProvider.loadConfig(configUrl)
    if (this.mounted) {
      this.setState({ config })
    }
  }

  private onGenerateHashes = async (profile: IProfile, inputIdentifier: string, inputToken: string, epochCount = 0): Promise<void> => {
    if (!this.state.config) {
      return
    }

    const input = `identifier=${inputIdentifier}&token=${inputToken}&epoch=${epochCount}`
    if (input === this.previousInput && profile === this.previousProfile) {
      return
    }
    this.previousInput = input
    this.previousProfile = profile

    console.log(`Generate hashes: with profile \'${profile.name}\' and identifier \'${inputIdentifier}\'`)
    const hashes = await HashBox.run(this.state.config.hashBoxConfig, profile, input)
    this.setState({ hashes })
  }
}

export default withStyles(styles)(App)
