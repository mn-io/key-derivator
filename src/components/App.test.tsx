import * as React from 'react'
import renderer from 'react-test-renderer'
import mock, { once } from 'xhr-mock'
import config from './../DefaultConfig'
import App from './App'

describe('', () => {

  beforeEach(() => mock.setup())

  afterEach(() => mock.teardown())

  it('renders without crashing', () => {
    mock.get('config.json', once({
      body: JSON.stringify(config),
      status: 200,
    }))

    const tree = renderer
      .create(<App />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

})
