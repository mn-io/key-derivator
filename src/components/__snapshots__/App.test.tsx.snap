// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<div
  className="App-root-3"
>
  <div
    className="App-row-4"
  >
    <div
      className="App-cell-1"
    >
      <div
        className="MuiPaper-root-6 MuiPaper-elevation2-10 MuiPaper-rounded-7 App-userInput-5"
      >
        <form
          onSubmit={[Function]}
        >
          <h5
            className="MuiTypography-root-37 MuiTypography-h5-53 MuiTypography-alignCenter-60 UserInput-spacingBottom-36"
          >
            key derivator
          </h5>
          <div
            className="MuiLinearProgress-root-73 MuiLinearProgress-colorPrimary-74 MuiLinearProgress-query-79"
            role="progressbar"
          >
            <div
              className="MuiLinearProgress-bar-83 MuiLinearProgress-barColorPrimary-84 MuiLinearProgress-bar1Indeterminate-86"
              style={Object {}}
            />
            <div
              className="MuiLinearProgress-bar-83 MuiLinearProgress-barColorPrimary-84 MuiLinearProgress-bar2Indeterminate-89"
              style={Object {}}
            />
          </div>
        </form>
      </div>
    </div>
    <div
      className="App-cell-1"
    />
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div
  className="App-root-3"
>
  <div
    className="App-row-4"
  >
    <div
      className="App-cell-1"
    >
      <div
        className="MuiPaper-root-6 MuiPaper-elevation2-10 MuiPaper-rounded-7 App-userInput-5"
      >
        <form
          onSubmit={[Function]}
        >
          <h5
            className="MuiTypography-root-37 MuiTypography-h5-53 MuiTypography-alignCenter-60 UserInput-spacingBottom-36"
          >
            key derivator
          </h5>
          <div
            className="MuiLinearProgress-root-73 MuiLinearProgress-colorPrimary-74 MuiLinearProgress-query-79"
            role="progressbar"
          >
            <div
              className="MuiLinearProgress-bar-83 MuiLinearProgress-barColorPrimary-84 MuiLinearProgress-bar1Indeterminate-86"
              style={Object {}}
            />
            <div
              className="MuiLinearProgress-bar-83 MuiLinearProgress-barColorPrimary-84 MuiLinearProgress-bar2Indeterminate-89"
              style={Object {}}
            />
          </div>
        </form>
      </div>
    </div>
    <div
      className="App-cell-1"
    />
  </div>
</div>
`;
